presto-main/src/main/java/com/facebook/presto/sql/FunctionInvoker.java

```diff
@@ -48,18 +51,33 @@ public Object invoke(Signature function, ConnectorSession session, Object... arg
      */
     public Object invoke(Signature function, ConnectorSession session, List<Object> arguments)
     {
-        MethodHandle method = registry.getScalarFunctionImplementation(function).getMethodHandle();
+        ScalarFunctionImplementation implementation = registry.getScalarFunctionImplementation(function);
+        MethodHandle method = implementation.getMethodHandle();
 
         List<Object> actualArguments = new ArrayList<>(arguments.size() + 1);
 
         Iterator<Object> iterator = arguments.iterator();
         for (int i = 0; i < method.type().parameterCount(); i++) {
-            if (method.type().parameterType(i) == ConnectorSession.class) {
+            Class<?> parameterType = method.type().parameterType(i);
+            if (parameterType == ConnectorSession.class) {
                 actualArguments.add(session);
             }
             else {
                 checkArgument(iterator.hasNext(), "Not enough arguments provided for method: %s", method.type());
-                actualArguments.add(iterator.next());
+                Object argument = iterator.next();
+                if (implementation.getNullFlags().get(i)) {
+                    boolean isNull = argument == null;
+                    if (isNull && !Primitives.isWrapperType(parameterType)) {
```

Comment:
The `Primitives.isWrapperType` call should not be needed. `Defaults.defaultValue` will return null for any non-primitive type, so it will just work out (because argument is already null in that case)